tests/index.test.js

```diff
@@ -0,0 +1,296 @@
+let analytics;
+let analyticModules;
+let mockedAnalyticModules;
+let modulesSchema;
+const testStr = 'test';
+const testObj = { test: 'test' };
+const moduleNames = ['mock0', 'mock1', 'mock2'];
+const supportedModuleNames = ['mock0', 'mock2'];
+const trackViewModule = { support: ['trackView'] };
+const trackEventModule = { support: ['trackEvent'] };
+const trackConversionModule = { support: ['trackConversion'] };
+const initializeModule = { support: 'initialize' };
+
+document.body.innerHTML = '<script></script>';
+Object.keys(__analytics).forEach((singleModule) => {
+  __analytics[singleModule].isEnabled = true;
+});
+
+const createMyModule = (support, name = 'mock') => {
+  const template = {
+    name,
+    trackView: jest.fn(),
+    trackEvent: jest.fn(),
+    trackConversion: jest.fn(),
+    initialize: () => new Promise((resolve) => {
+      setTimeout(() => {
+        resolve(testObj);
+      }, 300);
+    }),
+  };
+  return Object.keys(template).reduce((module, method) => {
+    if (support.includes(method) || method === 'name') {
+      return {
+        ...module,
+        [method]: template[method],
+      };
+    }
+    return module;
+  }, {});
+};
+const createModuleList = (moduleRules) => {
+  const modules = [];
+  moduleRules.forEach((moduleRule, i) => {
+    modules.push(createMyModule(moduleRule.support, `mock${i}`));
+  });
+  return modules;
+};
+const setAnalyticModulesMock = async (moduleRules) => {
+  jest.resetModules();
+  jest.unmock('../src/modules');
+  jest.doMock('../src/modules', () => ({ __esModule: true, default: createModuleList(moduleRules) }));
+  analyticModules = await import('../src/modules');
+  analytics = await import('../src/index');
+  // eslint-disable-next-line prefer-destructuring
+  mockedAnalyticModules = analyticModules.default;
+};
+
+describe('analytics function', () => {
+  describe('trackEvent function', () => {
+    beforeEach(async () => {
+      await setAnalyticModulesMock([trackEventModule]);
+    });
+    test('run trackEvent in module', async () => {
+      analytics.Analytics.trackEvent(testStr, testObj, moduleNames);
+      expect(mockedAnalyticModules[0].trackEvent).toHaveBeenCalledWith(testStr, testObj);
+    });
+    test('not run trackEvent when name of module is not passed', async () => {
+      analytics.Analytics.trackEvent(testStr, testObj, 'notThisModuleName');
+      expect(mockedAnalyticModules[0].trackEvent).not.toHaveBeenCalledWith(testStr, testObj);
+    });
+    test('run trackEvent in module with global properties as string', async () => {
+      analytics.Analytics.setGlobalProperties('global', 'asString');
+      analytics.Analytics.trackEvent(testStr, testObj, moduleNames);
+      expect(mockedAnalyticModules[0].trackEvent).toHaveBeenCalledWith(testStr, { global: 'asString', ...testObj });
+    });
+    test('run trackEvent in module with global properties as object', async () => {
+      analytics.Analytics.setGlobalProperties({ global: 'asObject' });
```

Comment:
Q: typeof 'asObject' is string 🤔  or did I miss something?

Reply:
Important is type of parameter passed to the setGlobalProperties function. In this case, it is object. I will rename the test to make it clear 😉

Reply:
🙈  fine I miss brackets, :thank_you: